sdk/go/src/sawtooth_sdk/client/encoder.go

```diff
@@ -0,0 +1,156 @@
+package client
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"sawtooth_sdk/protobuf/batch_pb2"
+	"sawtooth_sdk/protobuf/transaction_pb2"
+	"strings"
+	"time"
+)
+
+// See sdk/go/tests/client_test.go for example usage.
+
+type Transaction struct{ pb *transaction_pb2.Transaction }
+type Batch struct{ pb *batch_pb2.Batch }
+
+type TransactionEncoder struct {
+	privkey         []byte
+	pubkey          []byte
+	familyName      string
+	familyVersion   string
+	payloadEncoding string
+	addNonce        bool
+}
+
+// NewTransactionEncoder constructs a new encoder which can be used to generate
+// transactions and batches, and to serialize batches for submitting to the
+// REST API.
+func NewTransactionEncoder(privkey []byte, familyName, familyVersion, payloadEncoding string, addNonce bool) *TransactionEncoder {
+	return &TransactionEncoder{
+		privkey:         privkey[:],
+		pubkey:          GenPubKey(privkey),
+		familyName:      familyName,
+		familyVersion:   familyVersion,
+		payloadEncoding: payloadEncoding,
+		addNonce:        addNonce,
+	}
+}
+
+// NewTransaction Creates a new transaction and handles the construction and
+// signing of the transaction header.
+func (self *TransactionEncoder) NewTransaction(payload []byte, inputs, outputs, dependencies []string) Transaction {
+	hash := Encode(SHA512(payload), "hex")
+
+	header := &transaction_pb2.TransactionHeader{
+		BatcherPubkey:   Encode(self.pubkey, "hex"),
```

Comment:
This makes it impossible for Transaction creators and Batchers to be separate users (that's the whole reason we have a batcher pubkey). `BatcherPubkey` should be a settable default in the TransactionEncoder, which if not set (there or during transaction creation), should default to the signer's pubkey.